scipy/stats/morestats.py

```diff
@@ -2166,21 +2259,15 @@ def ansari(x, y):
     if repeats and (m < 55 or n < 55):
         warnings.warn("Ties preclude use of exact statistic.")
     if exact:
-        astart, a1, ifault = statlib.gscale(n, m)
-        ind = AB - astart
-        total = np.sum(a1, axis=0)
-        if ind < len(a1)/2.0:
-            cind = int(ceil(ind))
-            if ind == cind:
-                pval = 2.0 * np.sum(a1[:cind+1], axis=0) / total
-            else:
-                pval = 2.0 * np.sum(a1[:cind], axis=0) / total
```

Comment:
@tupui Typically I don't like dramatically refactoring old code if it's not necessary for the change at hand, however I found 2169-2183 confusing, and it was going to be more difficult to verify the correctness of the `alternatives` implementation if we tried to work it into this structure. Also, I think if/when we _do_ add an exact method that considers ties, at that point we would definitely want to move that (more complex) code out of the main function, and this prepares us for that. 
Do you agree that the new class `_ABW` is more clear and worth the refactoring now?

Reply:
Yes the new class seems appropriate. I was just unsure about the need of a class over functions here.

Reply:
Yes, that's part of the reason for it. It makes a bigger difference when recursive methods are used to get the null distribution and tables are built up that are relevant for inputs with other sizes, but still useful here.

Reply:
That's what I thought. Then perfect 👌